readme.md
# aws-unixtime

Proof of concept for leveraging Ansible to deploy an AWS CloudFormation stack containing a Lambda via a custom resource.

The purpose of the Lambda is to take an input of a timezone in the state/province format, and the output of the stack will return the specific timezone variable information

## Setup

### AWS Credentials

First thing is to setup your aws config in `~/.aws/config`:

```ini
[default]
region={{ region }}
output=json
aws_access_key_id={{ your_access_key_id }}
aws_secret_access_key={{ your_secret_access_key }}
```

See documentation for more info: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html

### Local Environment

Install the AWS CLI: https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html

Ensure `python3` is installed

```bash
git clone https://github.com/bwoolf1122/aws-unixtime.git
cd aws-unixtime
python3 -m venv venv/
source venv/bin/activate
pip install --upgrade pip
pip install -r requirements.txt
```

## Usage

Once your environnement is setup, you can now run the following commands:

### Deploy Stack

#### Variables

##### Default

```yaml
time_zone: "Europe/London"
time_zone_attrs: "unixtime"
```

##### Valid

Valid `time_zone_attrs` can be supplied in a comma seperated list:

```
*
abbreviation
client_ip
datetime
day_of_week
day_of_year
dst
dst_from
dst_offset
dst_until
raw_offset
timezone
unixtime
utc_datetime
utc_offset
week_number
```

Valid `time_zone`s can be found here: http://worldtimeapi.org/api/timezone

### Deploy Command

```bash
# use default vars
ansible-playbook ansible-build.yml

# example to specify specific attrs to return
ansible-playbook ansible-build.yml -e 'time_zone_attrs=unixtime,timezone'

# example to specify all attrs to return
ansible-playbook ansible-build.yml -e 'time_zone_attrs="*"'

# example to specify specific attrs to return and timezone
ansible-playbook ansible-build.yml -e 'time_zone_attrs=unixtime,timezone time_zone="Africa/Abidjan"'
```

Once this is run, `host_vars/localhost.yml` will be created with the stack and bucket name. NOTE: its best to leave this file be and let Ansible manage this

### Destroy Stack

The stack and bucket name will be read from `host_vars/localhost.yml`. __Once they have been destroyed, the local config information will be deleted__

```bash
ansible-playbook ansible-build.yml -e '{deploy: false, destroy: true}'
```


CHANGELOG.md
#Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.5]

### Removed
- `requirements-new.txt` - Removed unused file

## [1.4]

### Changes
- `ansible-build.yml` - Use `or` for set_fact when overwriting

## [1.3]

### Changes
- `readme.md` - Fix wording and add `cd` cmd

## [1.2]

### Changes
- `ansible-build.yml` - Fix stack and bucket var name include for override
- `requirement.txt` - Now include version numbers

### Added
- Added `CHANGELOG.md`
- Added `VERSION`

## [1.1]

### Changes
- add `host_vars/localhost.yml` to `.gitignore`

### Removed
- remove old build script


## [1.0]

### Added
- project ini


build.sh
#!/bin/bash
# Author Brandt Woolf
# Date: 2021-08-07
# Inspiration from: https://github.com/awsdocs/aws-lambda-developer-guide/tree/main/sample-apps/blank-python
set -eo pipefail

# generate s3 bucket
BUCKET_ID=$(dd if=/dev/random bs=8 count=1 2>/dev/null | od -An -tx1 | tr -d ' \t\n')
BUCKET_NAME=lambda-artifacts-$BUCKET_ID

# generate stack name
STACK_NAME=unixtime-stack-$BUCKET_ID

# record file with bucket name
echo $BUCKET_NAME > bucket-name.txt

# created s3 bucket
aws s3 mb s3://$BUCKET_NAME

# clear prioir builds
rm -rf package
cd function

# create deps
pip install --target ../package/python -r requirements.txt

cd ..

# upload deps to bucket # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-cli-package.html
aws cloudformation package --template-file template.yml --s3-bucket $BUCKET_NAME --output-template-file generated-s3-template.yml

# deploy stack with s3 path
aws cloudformation deploy --template-file generated-s3-template.yml --stack-name $STACK_NAME --capabilities CAPABILITY_NAMED_IAM


files/function/main.py
# Author: Brandt Woolf
# Date: 2021-08-07
# Inspiration from: https://towardsdatascience.com/extending-cloudformation-using-lambda-backed-custom-resources-c99b98cc7304

import requests
import logging
from crhelper import CfnResource

logger = logging.getLogger(__name__)

helper = CfnResource(
  json_logging=False,
  log_level='DEBUG',
  boto_level='CRITICAL'
)

def HttpGet(url):
  logger.info(f"Sending request to {url}")
  r = requests.get(url)
  if r.ok:
    logger.info(f"Status: {r.status_code} for URL: {url}")
    return r.text
  else:
    logger.error(f"Status: {r.status_code} for URL: {url}")
    r.raise_for_status()

def validate_attrs(attrs):
  valid_attrs = ['*', 'abbreviation','client_ip','datetime','day_of_week','day_of_year','dst','dst_from','dst_offset','dst_until','raw_offset','timezone','unixtime','utc_datetime','utc_offset','week_number']
  attrs = attrs.split(',')
  for attr in attrs:
    if attr not in valid_attrs:
      raise ValueError(f"{attr} not in {valid_attrs}")
  return attrs

def handler(event, context):
  helper(event, context)

@helper.create
def create(event, context):
  # log entry point and rest resource properties
  logger.info("Resource Created")
  ResourceProperties = event['ResourceProperties']

  # gather the requested zone
  zone = ResourceProperties['zone']
  logger.info(f"zone: {zone}")

  #gather the request attributes
  attrs = ResourceProperties['gather_attr']
  logger.info(f"attrs: {attrs}")

  validated_attrs = validate_attrs(attrs)
  logger.info(f"valid attrs: {validated_attrs}")

  logger.info("zone and valid attr(s) collected")

  # get list of valid zones
  zones = HttpGet('http://worldtimeapi.org/api/timezone')

  # check if requested zone is valid
  if zone not in zones:
    # raise error if zone not in http://worldtimeapi.org/api/timezone
    logger.error(f"zone {zone} not in {zones}")
    raise ValueError(f"zone {zone} not in {zones}")

  # get timezone info
  r = HttpGet(f"http://worldtimeapi.org/api/timezone/{zone}.txt")

  # parse timezone info into dict
  timezone = {}
  # split on newline first
  for line in r.split('\n'):
    # split on colon to get key value pair
    l = line.split(":")
    # created and assign new key pair
    timezone[l[0].strip()] = l[1].strip()
  # check for * to return all attrs
  if '*' in validated_attrs:
    for key in timezone:
      helper.Data[key] = timezone[key]
  # return the attrs included in validated_attrs
  else:
    for key in timezone:
      if key in validated_attrs:
        helper.Data[key] = timezone[key]
      else:
        helper.Data[key] = ''

@helper.update
def update(event, context):
  logger.info("Resource Updated")
  return "NewPhysicalResourceId"

@helper.delete
def delete(event, context):
  logger.info("Resource Deleted")
